Added Loss of Signal (LOS) detection of 💩💩💩 + iOS Support

durul opened the pull request:
Hi,

Here are the changes

I Refactored PissSocket and the UI to handle connection status. I updated PissActor to manage the connection state and notify AppState. I enhanced PissLabel to indicate connection status visually. I also added connection status text to the app's menu bar.
I Refactored PissSocket to improve its structure and naming conventions. I updated PissActor to PissSocket and changed the appState type to AppStateViewModel for consistency. I also added documentation comments for clarity, enhancing code readability and maintainability.

Jaennaet commented:
Baby's first PR 😍

Great that you cleaned things up, thank you! It'll probably be next week before I can actually test this myself, but there's one thing I can note right off the bat: unless the Lightstreamer Swift SDK is wildly different to how the Javascript one works, I think this doesn't actually handle *loss of signal* (LOS), ie. the loss of the ISS telemetry signal itself by whatever ground station(s) that receive it (which is what I meant in the README but was probably impossible to tell given how terse it was), but subscription updates from Lightstreamer that somehow got lost.

Now, if you don't feel like banging out the LOS logic too you can just eg. edit the PR title & README a bit to reflect 👆, buuut I'll just drop the LOS-related stuff here anyhow so it's written down somewhere, and if you (or anybody!) feel like taking a stab at this you've got something to start from at least. 

This PR is greatly appreciated in any case, so thank you ❤️

---

Sooo, the way eg. the [ISS-Mimic dashboard](https://github.com/ISS-Mimic/Mimic/blob/9733aaeb5e9f6a18ef73e2eb93821b7a7eab2350/docs/index.html) seems to handle LOS is that it has another subscription for receiving both the current timestamp on the ISS's side (used for detecting stale data, which we can maybe worry about Later™ unless, again, you or anybody want to take a stab at it), and the current status of the telemetry signal as a numerical value (and apparently anything but `24` means there's some variety of LOS.)

The ISS-Mimic dashboard also tells you _kind_ of LOS it is, but I don't think we care too much about that? Being able to differentiate between LOS and dropped updates from Lightstreamer would be great though.


Here's the "meat" of the LOS & stale signal detection stuff from that dashboard:

https://github.com/ISS-Mimic/Mimic/blame/9733aaeb5e9f6a18ef73e2eb93821b7a7eab2350/docs/index.html#L2605

```js
var timeSub = new Subscription('MERGE', 'TIME_000001', ['TimeStamp','Status.Class']);
```


https://github.com/ISS-Mimic/Mimic/blame/9733aaeb5e9f6a18ef73e2eb93821b7a7eab2350/docs/index.html#L3025-L3083


```js
client.subscribe(timeSub);


var statuses = {0:"Overridden",1:"Missing",2:"Dead",3:"Off-Scale High",4:"Off-Scale Low",5:"Out of Limit High",6:"Out of Limit Low",7:"Out of Calibration",8:"Overscaling",9:"Static",10:"SSME Controller",11:"SSME EIU",12:"Flight Critical MDM",13:"Payload MDM",14:"OI MDM",15:"OI DSC",16:"Lower Limit",17:"Upper Limit",18:"Event Limit",19:"Selmon Aggregate Measure",20:"Truncated",21:"User-Defined",22:"User-Defined",23:"User-Defined",24:"Default"}


timeSub.addListener({
  onItemUpdate: function (update) {
		  
        var unixtime = (new Date).getTime();
		var date = new Date(unixtime);
		var hours = date.getUTCHours();
		var minutes = "0" + date.getMinutes();
		var seconds = "0" + date.getSeconds();
		var timestmp = new Date().setFullYear(new Date().getFullYear(), 0, 1);
		var yearFirstDay = Math.floor(timestmp / 86400000);
		var today = Math.ceil((new Date().getTime()) / 86400000);
		var dayOfYear = today - yearFirstDay;


		var timestampnow = dayOfYear*24 + hours + minutes/60 + seconds/3600;
		//console.log("timestamp now: " + timestampnow);
  
  
  
		var status = update.getValue('Status.Class');
		var AOStimestamp = parseFloat(update.getValue('TimeStamp'));
		//console.log("Timestamp: " + update.getValue('TimeStamp'));
		var difference = timestampnow - AOStimestamp;
		var diff_seconds = difference*3600;
		console.log(diff_seconds);
		if (status === '24') 
		{
			document.getElementById("AOS").innerHTML = "Signal Acquired!";
			document.getElementById("AOS").bgColor = '#008000';
		}
		else 
		{
			document.getElementById("AOS").innerHTML = "Signal Lost " + status + " " + statuses[status];
			document.getElementById("AOS").bgColor = '#FF0000';
		}
		if(difference > 0.00153680542553047)
		{
			document.getElementById("AOS").innerHTML = "Stale Signal " + status + " " + statuses[status];
			document.getElementById("AOS").bgColor = '#FF8000';
		}
	  	if (diff_seconds < 0)
		{
			diff_seconds = diff_seconds+86399;	//Yes I am too lazy to account for the GMT day change
		}
		document.getElementById("telem_time").innerHTML = diff_seconds.toFixed(0) + " seconds ago";
		if (diff_seconds > 4) 
		{
			document.getElementById("telem_time").bgColor = '#FF8000';
		}
		else
		{
			document.getElementById("telem_time").bgColor = '#008000';
		}
	}
  });
});
```

durul commented:
> Baby's first PR 😍
> 
> Great that you cleaned things up, thank you! It'll probably be next week before I can actually test this myself, but there's one thing I can note right off the bat: unless the Lightstreamer Swift SDK is wildly different to how the Javascript one works, I think this doesn't actually handle _loss of signal_ (LOS), ie. the loss of the ISS telemetry signal itself by whatever ground station(s) that receive it (which is what I meant in the README but was probably impossible to tell given how terse it was), but subscription updates from Lightstreamer that somehow got lost.
> 
> Now, if you don't feel like banging out the LOS logic too you can just eg. edit the PR title & README a bit to reflect 👆, buuut I'll just drop the LOS-related stuff here anyhow so it's written down somewhere, and if you (or anybody!) feel like taking a stab at this you've got something to start from at least.
> 
> This PR is greatly appreciated in any case, so thank you once more ❤️
> 
> Sooo, the way eg. the [ISS-Mimic dashboard](https://github.com/ISS-Mimic/Mimic/blob/9733aaeb5e9f6a18ef73e2eb93821b7a7eab2350/docs/index.html) seems to handle LOS is that it has another subscription for receiving both the current timestamp on the ISS's side (used for detecting stale data, which we can maybe worry about Later™ unless, again, you or anybody want to take a stab at it), and the current status of the telemetry signal as a numerical value (and apparently anything but `24` means there's some variety of LOS.)
> 
> The ISS-Mimic dashboard also tells you _kind_ of LOS it is, but I don't think we care too much about that? Being able to differentiate between LOS and dropped updates from Lightstreamer would be great though.
> 
> Here's the "meat" of the LOS & stale signal detection stuff from that dashboard:
> 
> https://github.com/ISS-Mimic/Mimic/blame/9733aaeb5e9f6a18ef73e2eb93821b7a7eab2350/docs/index.html#L2605
> 
> ```js
> var timeSub = new Subscription('MERGE', 'TIME_000001', ['TimeStamp','Status.Class']);
> ```
> 
> https://github.com/ISS-Mimic/Mimic/blame/9733aaeb5e9f6a18ef73e2eb93821b7a7eab2350/docs/index.html#L3025-L3083
> 
> ```js
> client.subscribe(timeSub);
> 
> 
> var statuses = {0:"Overridden",1:"Missing",2:"Dead",3:"Off-Scale High",4:"Off-Scale Low",5:"Out of Limit High",6:"Out of Limit Low",7:"Out of Calibration",8:"Overscaling",9:"Static",10:"SSME Controller",11:"SSME EIU",12:"Flight Critical MDM",13:"Payload MDM",14:"OI MDM",15:"OI DSC",16:"Lower Limit",17:"Upper Limit",18:"Event Limit",19:"Selmon Aggregate Measure",20:"Truncated",21:"User-Defined",22:"User-Defined",23:"User-Defined",24:"Default"}
> 
> 
> timeSub.addListener({
>   onItemUpdate: function (update) {
> 		  
>         var unixtime = (new Date).getTime();
> 		var date = new Date(unixtime);
> 		var hours = date.getUTCHours();
> 		var minutes = "0" + date.getMinutes();
> 		var seconds = "0" + date.getSeconds();
> 		var timestmp = new Date().setFullYear(new Date().getFullYear(), 0, 1);
> 		var yearFirstDay = Math.floor(timestmp / 86400000);
> 		var today = Math.ceil((new Date().getTime()) / 86400000);
> 		var dayOfYear = today - yearFirstDay;
> 
> 
> 		var timestampnow = dayOfYear*24 + hours + minutes/60 + seconds/3600;
> 		//console.log("timestamp now: " + timestampnow);
>   
>   
>   
> 		var status = update.getValue('Status.Class');
> 		var AOStimestamp = parseFloat(update.getValue('TimeStamp'));
> 		//console.log("Timestamp: " + update.getValue('TimeStamp'));
> 		var difference = timestampnow - AOStimestamp;
> 		var diff_seconds = difference*3600;
> 		console.log(diff_seconds);
> 		if (status === '24') 
> 		{
> 			document.getElementById("AOS").innerHTML = "Signal Acquired!";
> 			document.getElementById("AOS").bgColor = '#008000';
> 		}
> 		else 
> 		{
> 			document.getElementById("AOS").innerHTML = "Signal Lost " + status + " " + statuses[status];
> 			document.getElementById("AOS").bgColor = '#FF0000';
> 		}
> 		if(difference > 0.00153680542553047)
> 		{
> 			document.getElementById("AOS").innerHTML = "Stale Signal " + status + " " + statuses[status];
> 			document.getElementById("AOS").bgColor = '#FF8000';
> 		}
> 	  	if (diff_seconds < 0)
> 		{
> 			diff_seconds = diff_seconds+86399;	//Yes I am too lazy to account for the GMT day change
> 		}
> 		document.getElementById("telem_time").innerHTML = diff_seconds.toFixed(0) + " seconds ago";
> 		if (diff_seconds > 4) 
> 		{
> 			document.getElementById("telem_time").bgColor = '#FF8000';
> 		}
> 		else
> 		{
> 			document.getElementById("telem_time").bgColor = '#008000';
> 		}
> 	}
>   });
> });
> ```

Hahah from Baby's 😂 😂 😂

I added these changes:
1. Add signal status tracking via a new subscription to TIME_000001
2. Differentiate between connection loss (Lightstreamer issues) and signal loss (ISS telemetry issues)
3. Show different status messages and colors for each state
4. Use orange for LOS (like ISS-Mimic) and red for connection loss

durul commented:
![Simulator Screenshot - iPhone 16 - 2024-12-25 at 13 33 29](https://github.com/user-attachments/assets/b5f96dba-3823-4cef-a227-95070cc12e18)

durul commented:
![Simulator Screenshot - iPhone 16 - 2024-12-25 at 13 51 54](https://github.com/user-attachments/assets/105783c8-359c-492c-a6a8-673f050423f1)

durul commented:
Merry Xmas

Jaennaet commented:
Here, you dropped this, king: 👑

OK this is such an epic PR that I'm going to get this merged and released today, screw the holidays.

@durul, I was thinking of adding a "contributors" section to the README since it's not just me anymore, you OK with having your @ on it? If not that's totally fine too.

Oh, and merry xmas to you too! Thank you for the great work, you clearly know what you're doing with Swift & SwiftUI much better than I do 😀

Jaennaet commented on pISSStream/View/PissLabel.swift:
```diff
@@ -0,0 +1,28 @@
+    
+    var body: some View {
+        HStack(spacing: 4) {
+            Circle()
+                .fill(isConnected ? Color.green : Color.red)
+                .frame(width: 8, height: 8)
```
This doesn't show up on my 14.7.1 system:

![Screenshot](https://github.com/user-attachments/assets/6cc4873b-9837-4580-8252-c5b7095a5cff)

Should this work on eg. macOS 15.x?

Jaennaet replied:
Eh you know what, I'm just going to go ahead and merge this, maybe with a change so that the emojis change to eg. 🧑🏽‍🚀❗ on LOS. We can worry about details later 😀

durul replied:
> This doesn't show up on my 14.7.1 system:
> 
> ![Screenshot](https://private-user-images.githubusercontent.com/191785648/398599157-6cc4873b-9837-4580-8252-c5b7095a5cff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxNjc5ODEsIm5iZiI6MTczNTE2NzY4MSwicGF0aCI6Ii8xOTE3ODU2NDgvMzk4NTk5MTU3LTZjYzQ4NzNiLTk4MzctNDU4MC04MjUyLWM1YjcwOTVhNWNmZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNVQyMzAxMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMzEyN2ViMjZhODFjM2JmMTVkYjNhNjIxNDUzNjc0NzEyZmIwMjYxMDZmMDkyNDZlZTlkODJlNzY0YTAyMjBjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.AUIyKKwswzdUHMlgfO7z3n_50Ie9dyrwiWfNwgYezs8)
> 
> Should this work on eg. macOS 15.x?

Yes I am using 15.2 (24C101). Here is my screenshot

![Screenshot 2024-12-25 at 5 04 39 PM](https://github.com/user-attachments/assets/0fad0ac4-efa3-4ef6-a4c6-acbbf97e7ba0)

durul replied:
another one 

![Screenshot 2024-12-25 at 5 08 02 PM](https://github.com/user-attachments/assets/fa3f7d73-88e7-4b9d-b944-691e9a89e64d)

durul commented:
> Here, you dropped this, king: 👑
> 
> OK this is such an epic PR that I'm going to get this merged and released today, screw the holidays.
> 
> @durul, I was thinking of adding a "contributors" section to the README since it's not just me anymore, you OK with having your @ on it? If not that's totally fine too.
> 
> Oh, and merry xmas to you too! Thank you for the great work, you clearly know what you're doing with Swift & SwiftUI much better than I do 😀

#### Thank you so much! I’m thrilled to have contributed to this project, and I’d be honored to be included in the “contributors” section of the README.
#### 
Also, It’s been a pleasure collaborating, and I’m looking forward to seeing how this project evolves. Let me know if there’s anything else I can help with! 😊🎄
